python/nav/web/webfront/__init__.py

```diff
@@ -23,17 +23,41 @@ def find_dashboard(account, dashboard_id=None):
     Either find a specific one or the default one. If none of those exist we
     find the one with the most widgets.
     """
-    kwargs = {'pk': dashboard_id} if dashboard_id else {'is_default': True}
+    dashboard = (
+        _find_dashboard_by_id(account, dashboard_id)
+        if dashboard_id
+        else _find_default_dashboard(account)
+    )
+    dashboard.shared_by_other = dashboard.is_shared and dashboard.account != account
+    dashboard.is_default = dashboard.is_default_for_account(account)
+
+    return dashboard
+
+
+def _find_dashboard_by_id(account, dashboard_id):
+    """Find a specific dashboard by ID for this account"""
+    kwargs = {'pk': dashboard_id}
     try:
         dashboard = AccountDashboard.objects.get(
             (Q(account=account) | Q(is_shared=True)), **kwargs
         )
-        dashboard.shared_by_other = dashboard.is_shared and dashboard.account != account
+        return dashboard
 
     except AccountDashboard.DoesNotExist:
-        if dashboard_id:
-            raise Http404
+        raise Http404
+
 
+def _find_default_dashboard(account):
+    """Find the default dashboard for this account"""
+    kwargs = (
+        {'pk': account.default_dashboard.id} if account.has_default_dashboard else {}
+    )
+    try:
+        dashboard = AccountDashboard.objects.get(
+            (Q(account=account) | Q(is_shared=True)), **kwargs
+        )
+
+    except AccountDashboard.DoesNotExist:
```

Comment:
This feels quite convoluted, why are we not using `account.shared_dashboard` here and then the exception does everything after `except AccountDashboard.DoesNotExist:`? 

Or do a general `if account.has_default_dashboard` check and then in else part we check if there exists at least one dashboard and then return at the one with the most widgets